Monaco

bcolloran opened the pull request:
This may now be working at parity with codemirror, but i need extra eyes and review to kick the tires. I guess the one thing that is for sure not working is the fetch chunk code highlighting mode. i can look at that next.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Documentation**: If this feature has or requires documentation, the relevant docs have been updated.
- [ ] **Changelog**: This PR updates the [changelog](../CHANGELOG.md) with any user-visible changes.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not

bcolloran commented:
fetch chunk highlighting now working. last explicit bug i'm now aware of: chunk delim lines without an explicit type don't highlight correctly.

@hamilton @mdboom @wlach we should figure out what the parity-with-codemirror criteria are for Monaco. Getting all the magic of language servers working is still a ways off, but i think i can see a path to some cool stuff that is pretty attainable

wlach commented on src/editor/components/monaco-language-init.js:
```diff
@@ -0,0 +1,39 @@
+  md: [mdLang, mdConf],
+  css: [cssLang, cssConf],
+  py: [pyLang, pyConf]
+}).forEach(lang => {
```
Do we need to update the language plugin documentation for any new conventions here?

bcolloran commented:
@hamilton @mdboom @wlach @teonbrooks -- i've made some changes based on your feedback, primarly WRT autocomplete. I added some goodies for iomd (chunk type and chunk flag completion, and fetch chunk completions) but the main thing is that now code-ish chunk that might share variable names share completion tokens (css, js, py, fetch), and md chunks use tokens from the entire document (in case you want to reference a variable name from your code in your MD?). i'm really not at all tied to this approach and happy to change it based on feedback, it's just kind of an experiment, but i thought that maybe code chunks (even of different languages) would likely share identifier strings.

so again, please kick the tires and let me know what you think. press `ctrl+space` liberally :-)

also, the code is still WIP, so please just focus on QA ing the functionality -- when we know what we want i'll tighten things up for a proper review pass.

also @wlach (an anyone else who has an opinion) the wider gutter is used by default for monaco for code folding icon button thingies and i think other gutter indicators. i don't know how much people actually use code folding, but i think some of the other gutter content might be relevant when we eventually get intellisense stuff working. but i can look into shrinking that down more if anyone has strong feelings

bcolloran commented:
(oh, also removed the minimap)

wlach commented:
I deployed this to https://definitely-staging.iodide.io in case anyone wants to give it a whirl. Will give actual feedback tomorrow.

teonbrooks commented:
so far planning around with the editor, i think the languages shared refs make sense. i understand the inclusion of the fetch cells as well, that is the loaded data gets its variable assignment there, but what i have noticed with this is that it parts of the url are included in the autocomplete list.

bcolloran commented:
@teonbrooks -- ok i fooled around with it a bit more, and tried to tighten things up a bit. Since I already had the regex handy from something else, i added a filter so that _only_ the variable names in fetch chunks are added to the autocomplete list, and i added a rudimentary heuristic so that only things in css chunks starting with "#" or "." (ids and classes) are added to the list. please take another look when you can and let me know if you see any other merge blockers.

thanks for helping with the QA @teonbrooks !

bcolloran commented on webpack.config.js:
```diff
@@ -141,7 +138,6 @@ module.exports = env => {
         filename: `[name].${APP_VERSION_STRING}.css`
       }),
       new WriteFilePlugin()
-      // Use an external helper script, due to https://github.com/1337programming/webpack-shell-plugin/issues/41
```
comment for code that was removed

bcolloran commented on webpack.config.js:
```diff
@@ -2,10 +2,8 @@ require("dotenv").config();
 const webpack = require("webpack");
 const path = require("path");
 const fs = require("fs");
-const CreateFileWebpack = require("create-file-webpack");
```
this and the one below were no longer used. noticed this while cleaning out stuff i'd been doing in this file.

bcolloran commented on src/editor/state-schemas/language-definitions.js:
```diff
@@ -19,8 +16,6 @@ const PYODIDE_URL = process.env.USE_LOCAL_PYODIDE
 const pyLanguageDefinition = {
   languageId: "py",
   displayName: "Python",
-  codeMirrorMode: "python",
```
i don't think it should be hard, but i think we can address that down the line. there will be a bit more thinking to do on this topic, but not too bad.

teonbrooks commented:
@bcolloran that did it for me. looks good!

bcolloran commented:
note to self: WIP of webpack.config.js with stuff for loading monaco here:
https://gist.github.com/bcolloran/9ad3686b793795bcd03ae285aa24b8e1

wlach commented on src/editor/components/monaco-tools/__tests__/code-chunk-identifiers.test.js:
```diff
@@ -0,0 +1,114 @@
+import { fetchIdentifiers, cssIdentifiers } from "../code-chunk-identifiers";
```
I wonder if it might make more sense to just have an `iomd-editor` directory to just put all this stuff in:

`src/editor/components/iomd-editor/` 

Then could just dump all these extra js files in there, along with `iomd-editor.jsx`. I don't think it would be necessary to say these are "monaco-tools" in the directory structure then.

bcolloran replied:
yep good idea

wlach commented on src/editor/components/monaco-tools/delim-line-suggestion.js:
```diff
@@ -0,0 +1,40 @@
+// be available at run time. but for now, we can just hardcode this here.
+export const validChunkFlags = ["skipRunAll"];
+
+export const delimLineSuggestion = (lineSoFar, knownChunkTypes, lineNumber) => {
```
Would be nice to have a jest test for this function.

bcolloran replied:
added tests! :-)

wlach commented on src/editor/components/monaco-tools/fetch-line-suggestion.js:
```diff
@@ -0,0 +1,73 @@
+
+const { Keyword, Snippet, File } = monaco.languages.CompletionItemKind;
+
+export const fetchLineSuggestion = (lineSoFar, fileNames) => {
```
Likewise, a test for this function sounds like a good idea, at least covering a few basic cases.

bcolloran replied:
added tests! :-)

bcolloran commented on src/editor/components/monaco-tools/code-chunk-identifiers.js:
```diff
@@ -0,0 +1,55 @@
+  // FIXME: this implementation is not terribly clever or efficient.
+  // If we move this to the store or a language worker, perhaps
+  // we can more intelligently manage the list of words incrementally
+  // rather than rebuilding each time. Or ideally, we can use an existing tokenizer
```
ya exactly, jank is the worry here. monaco recalculates everything as you type, so ideally we'd tap into it's tokenizer, but there is apparently no stable API for accessing it. however, this function seems to be performing acceptably for the time being, even on unusually large test notebooks

bcolloran commented on patches/monaco-editor+0.16.2.patch:
```diff
@@ -0,0 +1,43 @@
++            if (browser.isFirefox) {
++                this._register(dom.addDisposableListener(this._listenOnDomNode, 'DOMMouseScroll', onMouseWheel, true));
++            } else {
++                this._register(dom.addDisposableListener(this._listenOnDomNode, 'mousewheel', onMouseWheel, true));
```
yep, necessary. the next monaco update should have a fix that allows us to remove this.

bcolloran commented on src/editor/components/iomd-editor.jsx:
```diff
@@ -3,227 +3,219 @@ import { connect } from "react-redux";
   static propTypes = {
     content: PropTypes.string,
-    editorOptions: PropTypes.object, // eslint-disable-line react/forbid-prop-types
+    wordWrap: PropTypes.string.isRequired,
```
yep np

bcolloran commented on src/editor/components/iomd-editor.jsx:
```diff
@@ -3,227 +3,219 @@ import { connect } from "react-redux";
   constructor(props) {
     super(props);
+    this.containerDivRef = React.createRef();
+    this.editor = null;
```
nope, i suppose i do not :-)

bcolloran commented on src/editor/components/monaco-tools/delim-line-suggestion.js:
```diff
@@ -0,0 +1,40 @@
+    suggestions = makeSuggestionList(
+      knownChunkTypes.map(ct => `${"%".repeat(numPctSigns)} ${ct}`),
+      Keyword,
+      { range: new monaco.Range(lineNumber, 1, lineNumber, numPctSigns) }
```
good idea

bcolloran commented on src/editor/components/iomd-editor.jsx:
```diff
@@ -3,227 +3,219 @@ import { connect } from "react-redux";
-    keyMap: "sublime",
-    comment: true
-  };
+  const wordWrap = state.wrapEditors ? "on" : "off";
```
i think mapStateToProps is for translating data in the redux store to the data needed for a component so that you can have less logic in the component

wlach replied:
In general I agree, though the fact that monaco expects a parameter with a string argument "on"/"off" feels like an implementation detail that is best dealt with inside the thing that's talking to monaco.

bcolloran commented on src/editor/components/iomd-editor.jsx:
```diff
@@ -3,227 +3,219 @@ import { connect } from "react-redux";
+  // works great.
+  const editorPositionString = Object.values(
+    state.panePositions.EditorPositioner
+  ).join(",");
```
ya i know, <sheepish> but this actually kind of makes things easier WRT proptypes etc, and it's less code running behind the scenes and it's one less import.

all that said, i can change it if you think it's ugly

wlach replied:
It's just an array of numbers, no? Should be pretty easy to represent in proptypes.

My counter-argument would be that converting to a string just to do a comparison is inefficient (though obviously this function isn't being called enough for that to remotely matter in 2019) and could mask errors in certain cases (which admittedly are quite unlikely).

Probably not worth arguing about more than that. I'd encourage you to give my suggested approach a try (I think it'll take less than 5 minutes) but won't pursue the point further if you really want to do it this way.

bcolloran replied:
no you're right. it's an easy fix. thank you for keeping me honest :-)
